packages/SwingSet/src/kernel/virtualObjectManager.js

```diff
@@ -0,0 +1,342 @@
+        } else {
+          assertKeyExists(vkey);
+          backingMap.set(key, value);
+        }
+      },
+    });
+  }
+
+  /**
+   * Fetch an object's state from secondary storage.
+   *
+   * @param {string} instanceKey  The instance ID of the object whose state is
+   *    being fetched.
+   *
+   * @returns {*} an object representing the object's stored state.
+   */
+  function fetch(instanceKey) {
+    return JSON.parse(syscall.vatstoreGet(instanceKey));
+  }
+
+  /**
+   * Write an object's state to secondary storage.
+   *
+   * @param {string} instanceKey  The instance ID of the object whose state is
+   *    being stored.
+   * @param {*} rawData  A data object representing the state to be written.
+   */
+  function store(instanceKey, rawData) {
+    syscall.vatstoreSet(instanceKey, JSON.stringify(rawData));
+  }
+
+  /**
+   * Make a new kind of virtual object.
+   *
+   * @param {*} instanceMaker  A function of the form `instanceMaker(state)` that
+   *    will return a representative instance wrapped around the given state.
+   *
+   * @returns {*} a maker function that can be called to manufacture new
+   *    instance of this kind of object.  The parameters of the maker function
+   *    are those of the `initialize` method implemented in the representative
+   *    produced by the `instanceMaker` parameter.
+   */
+  function makeKind(instanceMaker) {
+    const kindID = `${allocateExportID()}`;
+    let nextInstanceID = 1;
+
+    function makeInnerSelf(instanceKey, rawData) {
+      const innerSelf = {
+        prev: undefined,
+        next: undefined,
+        instanceKey,
+        rawData,
+        dropData() {
+          store(instanceKey, innerSelf.rawData);
+          innerSelf.rawData = null;
+        },
+      };
+      return innerSelf;
+    }
+
+    function makeRepresentative(innerSelf, initArgs) {
+      function ensureState() {
+        if (!innerSelf.rawData) {
+          const instanceKey = innerSelf.instanceKey;
+          innerSelf = cache.lookup(instanceKey);
+          if (!innerSelf) {
+            innerSelf = makeInnerSelf(instanceKey, fetch(instanceKey));
+            cache.remember(innerSelf);
+          }
+        }
+        cache.refresh(innerSelf);
+      }
+
+      function wrapData() {
+        innerSelf.activeData = {};
+        for (const prop of Object.getOwnPropertyNames(innerSelf.rawData)) {
+          Object.defineProperty(innerSelf.activeData, prop, {
+            get: () => {
+              ensureState();
+              return m.unserialize(innerSelf.rawData[prop]);
+            },
+            set: value => {
+              ensureState();
+              innerSelf.rawData[prop] = m.serialize(value);
+            },
+          });
+        }
+        return harden(innerSelf.activeData);
+      }
+
+      if (!innerSelf.rawData && initArgs) {
+        innerSelf.rawData = {};
+        const tempInstance = instanceMaker(innerSelf.rawData);
```

Comment:
This weirded me out too: sometimes `rawData` is used for the behavior object's `state` variable, but sometimes it's used to hold the serialized data that the getters/setters manage.

Reply:
Yup.  The alternative is having the initializer be a separate parameter the way MarkM advocates, which I see as a major user ergonomics disimprovement solely motivated by considerations of internal abstract purity.  The key thing in my view is that from the perspective of the implementor of a new kind of object (i.e., of the code being handed to `makeKind`) `state` is always just the state and the behind-the-scenes magic is transparent.

Reply:
In this case, I think we could use some other object instead of `innerSelf`, since the temporary instance isn't sticking around very long, and the object we give it to hold data won't disappear before we're done with `initialize`. See my sketch for an alternative.

Reply:
Without going into your alternative (which I haven't completely internalized yet), if I understand correctly this would basically be a rewrite from
```javascript
        innerSelf.rawData = {};
        const tempInstance = instanceMaker(innerSelf.rawData);
        tempInstance.initialize(...initArgs);
        for (const prop of Object.getOwnPropertyNames(innerSelf.rawData)) {
          innerSelf.rawData[prop] = m.serialize(innerSelf.rawData[prop]);
        }
```
to
```javascript
        const initializationData = {};
        const tempInstance = instanceMaker(initializationData);
        tempInstance.initialize(...initArgs);
        innerSelf.rawData = {};
        for (const prop of Object.getOwnPropertyNames(initializationData)) {
          innerSelf.rawData[prop] = m.serialize(initializationData[prop]);
        }
```
the payoff being that it would preclude the possibility of certain later nonsense if `initialize` were to squirrel away its parameter so somebody could do something with it in the future.  Minor refactor, clear win.

Reply:
Done.

Reply:
yeah, plus then `rawData` is only ever holding the serialized form, instead of (briefly) holding the non-serialized form during initialization, which I think makes it less confusing to read